### ISSUE TITLE
Overriding tidyverse function arg style

### ISSUE THREAD

**Issue by JosiahParry** [r-lib/styler#1184]
I'd pose this in a discussion but there are not any for this repo!

The tidyverse style guide says: 

> In both cases the trailing ) and leading { should go on the same line as the last argument.

However, it the closing `)` and opening `{` on the same line as the last argument give me the ick! 😜 Is there a way to modify styler to put the closing paren and opening bracket on their own line when there are many function arguments? 

I use [codegrip](https://github.com/lionel-/codegrip) to format my function arguments. However, I have recently configured my environment to use styler to format on save. This one styling, though, i'd like to override. 

```r
reverse_geocode <- function(
    locations,
    crs = sf::st_crs(locations),
    ...,
    lang_code = NULL,
    feature_type = NULL,
    for_storage = FALSE,
    location_type = c("rooftop", "street"),
    preferred_label_values = c("postalCity", "localCity"),
    geocoder = default_geocoder(),
    token = arc_token(),
    .progress = TRUE
) {
```

**Comment by IndrajeetPatil** [comment]
I agree that this should be the default, but, unfortunately, since the underlying style guide specifies this behaviour, we need to stick to it. 

As for `codegrip`, there is a [PR](https://github.com/lionel-/codegrip/pull/16) to bring it in line with the tidyverse style guide, so it might also start behaving the same way as `{styler}`.

That said, [`{grkstyle}`](https://github.com/gadenbuie/grkstyle) defaults to the behaviour you desire, so you might want to check it out.

**Comment by IndrajeetPatil** [comment]
It's not `{styler}` that is opinionated, it's the style guide 🙃

But `{styler}` does allow further customizations: https://styler.r-lib.org/articles/customizing_styler.html

TBH, I'd much rather the style guide change its recommendation here.

**Comment by JosiahParry** [comment]
`{styler}` is enforcing the style guide, thus opinionated! 🙈 
Are the stylers documented? It's tough to figure out which one it might be.

Say I've figured out how to customize my styler, how do I utilize it as a global standard? For example with lintr, there is a `.lintr` file. The customizing-style, third party integration, and distribute custom styles vignettes do not describe how to make a custom styler the default when working with styler. 

For context, I'm using VS Code to format on save and not manually styling a file from the R console. Ideally it can be modified in the Rprofile or other environment location

**Comment by JosiahParry** [comment]
I believe this may be possible in https://github.com/r-lib/lintr/pull/1411 by @MichaelChirico but unclear how 🤔

**Comment by lorenzwalthert** [comment]
> The customizing-style, third party integration, and distribute custom styles vignettes do not describe how to make a custom styler the default when working with styler.

You don't use {styler} directly anymore if you create your own style guide based on the styler infra, you just call your own 
pkg. See this [README](https://github.com/lorenzwalthert/styler.nocomments).
```r
library(styler.noncomments)
style_text('1 # comment') # this default to transformers = nocomments_style
```

**Comment by lorenzwalthert** [comment]
a directory specific configuration for styler is an open issue: #319

**Comment by JosiahParry** [comment]
@lorenzwalthert I have no intention of creating and distributing my own R package for modifying a single style rule. It would ideal to be able to change or remove one rule in the transformers list and have that be used. Perhaps it can be set via `options()` in a user-level `.Rprofile`?

**Comment by lorenzwalthert** [comment]
For VS code specifically, see https://github.com/REditorSupport/languageserver?tab=readme-ov-file#customizing-formatting-style

**Comment by lorenzwalthert** [comment]
I think it seems like the style guide might change in the future or at least allow both options since Lionel supports your point of view in https://github.com/lionel-/codegrip/pull/16, but for this to happen, I think as @IndrajeetPatil said, it needs a formal decision upstream. Depending on what happens there, we could change the behaviour or expose an argument in `tidyverse_style()` to configure the behaviour.

**Comment by lorenzwalthert** [comment]
> @lorenzwalthert I have no intention of creating and distributing my own R package for modifying a single style rule. 

Then I am afraid you have to adhere to the style guide, or open an issue in tidyverse/style.

**Comment by JosiahParry** [comment]
Heard. 

Just to be crystal clear, the maintainers of `{styler}` have no intention of making it possible to simply modify styler similar to `lintr`'s ability to modify them via options e.g.

```r
options(lintr.line_length_lintr = NULL)
```

and instead, require users to create their own R package for this behavior?

**Comment by lorenzwalthert** [comment]
I might be able to show you how you can modify styler to achieve what you want, but unless you create your own style guide, the option won’t be as portable as a simple one liner in a co fig file like for linter.

**Comment by lorenzwalthert** [comment]
Intentions we have, but time and resources and priorities are a different topic 😜

**Comment by JosiahParry** [comment]
Closing in favor of #319

**Comment by lionel-** [comment]
FWIW I think we'd be supportive of a style guide PR to bring it in line with codegrip behaviour. We've originally aligned our guide to the google C++ one, but after working with Rust and Typescript where the balanced style seem to be the norm we think it's best to align with these languages.

**Comment by JosiahParry** [comment]
This is now official tidyverse style see https://style.tidyverse.org/functions.html#multi-line-function-definitions

Can we consider reopening this issue?

**Comment by IndrajeetPatil** [comment]
With the latest commit on the `main`-branch:

``` r
"reverse_geocode <- function(
  locations,
  .progress = TRUE) {
  NULL
}" -> code

styler::style_text(code)
#> reverse_geocode <- function(
#>   locations,
#>   .progress = TRUE
#> ) {
#>   NULL
#> }
```

<sup>Created on 2024-11-25 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by lorenzwalthert** [comment]
I guess we can close this now.

**Comment by IndrajeetPatil** [comment]
@lorenzwalthert Should we create a new CRAN release? 

I would say that enough has changed to make it available to the users.

**Comment by lorenzwalthert** [comment]
I think we should also implement some of the other open issues that resulted from the style guide update before that.

**Comment by IndrajeetPatil** [comment]
@lorenzwalthert So only after closing this issue?

https://github.com/r-lib/styler/issues/1231

One of those items is open upstream, so it might be a while before that issue can be closed. But maybe we can try to address the other two items in it and prepare a release.
